lib/Dataflow_analysis.ml

```diff
@@ -0,0 +1,800 @@
+
+(**
+   A label is a unique identifier for a node in the dataflow/dependency graph, and
+   usually corresponds to one node in the Mir.
+*)
+type label = int [@@deriving sexp]
+
+(**
+   A 'reaching dependency' (or reaching_dep or RD) statement (v, l) says that the variable
+   v could have been affected at the label l.
+*)
+type reaching_dep = expr * int [@@deriving sexp, hash, compare]
+
+(**
+   A reaching dependency set holds the set of reaching dependencies that could be true at
+   some point.
+*)
+module ReachingDepSet = Set.Make (struct
+  type t = reaching_dep
+
+  let compare : reaching_dep -> reaching_dep -> int = compare
+  let sexp_of_t = sexp_of_reaching_dep
+  let t_of_sexp = reaching_dep_of_sexp
+end)
+
+module LabelMap = Map.Make (struct
+  type t = label
+
+  let compare : int -> int -> int = compare
+  let sexp_of_t = sexp_of_int
+  let t_of_sexp = int_of_sexp
+end)
+
+module LabelSet = Set.Make (struct
+  type t = label
+
+  let compare : int -> int -> int = compare
+  let sexp_of_t = sexp_of_int
+  let t_of_sexp = int_of_sexp
+end)
+
+module ExprSet = Set.Make (struct
+  type t = expr
+
+  let compare : expr -> expr -> int = compare
+  let sexp_of_t = sexp_of_expr
+  let t_of_sexp = expr_of_sexp
+end)
+
+(**
+   Description of where a node in the dependency graph came from, where MirNode is the
+   location from an Mir.loc_stmt
+ *)
+type source_loc =
+  | MirNode of string
+  | StartOfBlock
+  | TargetTerm of {term: expr; assignment_label: label}
+[@@deriving sexp]
+
+(**
+   Information to be collected about each node
+   * dep_sets: Information about how the label effects the dependency set
+   * possible_previous: The set of nodes that could have immediately preceded this node
+     under some execution of the program
+   * rhs_set: The 'right hand side' set of variables that affect the value or behavior of
+     this node
+   * controlflow: The set of control flow nodes that are immediate parents of this node:
+     * The most recent nested if/then or loop,
+     * or the beginning of the function or block if there are no containing branches,
+     * plus the set of relevant continue/return statements,
+     * plus, for loops, any break statements they contain
+   * loc: The location of the Mir node that this node corresponds to, or a description if
+     there is none
+*)
+type 'dep node_info =
+  { dep_sets: 'dep
+  ; possible_previous: LabelSet.t
+  ; rhs_set: ExprSet.t
+  ; controlflow: LabelSet.t
+  ; loc: source_loc }
+[@@deriving sexp]
+
+(**
+   A node_info, where the reaching dependency information takes the form of an update
+   function that maps from the 'entry' set to the 'exit' set, where the entry set is
+   what's true before executing this node and the exit set is true after.
+*)
+type node_info_update = (ReachingDepSet.t -> ReachingDepSet.t) node_info
+
+(**
+   A node_info where the reaching dependency information is explicitly written as the
+   entry and exit sets, as after fixpoint analysis.
```

Comment:
what is "fixpoint analysis?" do you mean after finding the fixed point? ie after running the analysis over and over and seeing nothing change?

Reply:
Right. It's the terminology used in the textbook Nielson & Nielson. Apply the RD update functions repeatedly until the sets stop changing. They have a proof of convergence in the book.

Reply:
Ah, I stopped reading that book after I read [this review](https://www.amazon.com/gp/customer-reviews/R2G1TETI817U5W/ref=cm_cr_dp_d_rvw_ttl?ie=UTF8&ASIN=3540654100) :P I don't see the phrase "fixpoint analysis" but I'm willing to believe they use it. I think I'm just not used to seeing fixpoint used 1) instead of "fixed-point" or 2) as an adjective. I wonder if it would be easier to read as something like 
```
A node_info where the reaching dependency information is explicitly written as the
   entry and exit sets, as obtained after running the analysis until finding this fixed-point.
```

Reply:
You're right, fixed-point is better, fixpoint is just the shorthand I use in my head.